doc/src/modules/solvers/solveset.rst

```diff
@@ -1,12 +1,463 @@
+
+ Though, there still a lot of work needs to be done in this regard.
+
+
+Search Based solver & Step by Step Solution
+-------------------------------------------
+
+ Note: This is under Development.
+
+ After the introduction of ``ConditionSet`` [10], the solving of equations can be
+ seen as set transformation, We can do the following things to solve equations
+ (abstract View):
+
+ * Apply Various Set Transformations on the given Set.
+ * Define a Metric of the usability or define a notion of better solution over others.
+ * Different Transformation would be the nodes of the tree.
+ * Suitable searching techniques could be applied to get the best solution.
+
+ ConditionSet gives us the ability to represent unevaluated equations and
+ inequalities in forms like `{x|f(x)=0; x in S}` and `{x|f(x)>0; x in S}`
+ but a more powerful thing about ConditionSet is that it allows us to write
+ the intermediate steps as set to set transformation. Some of the transformations
+ are:
+
+ * Composition: `{x|f(g(x))=0;x in S} => {x|g(x)=y; x in S,y in {z|f(z)=0; z in S}}`
+
+ * Polynomial Solver: `{x|P(x)=0;x in S} => {x_1,x_2, ... ,x_n}.intersection(S)`
+   where `x_i` are roots of P(x)
+
+ * Invert solver: `{x|f(x)=0;x in S} => {g(0)| all g such that f(g(x)) = x}`
+
+ * logcombine: `{x| log(f(x)) + log(g(x));x in S}`
+                => `{x| log(f(x)*g(x)); x in S}` if f(x) > 0 and g(x) > 0
+                => `{x| log(f(x)) + log(g(x));x in S}` otherwise
+
+ * product solve: {x|f(x)*g(x)=0; x in S}
+                => `{x|f(x)=0; x in S} U {x|g(x)=0; x in S}` given f(x) and g(x) are bounded
+                => {x|f(x)*g(x)=0; x in S}, otherwise
+
+ Since the output type is same as input type any composition of these
+ transformations is also a valid transformation. And our aim is to find
+ the right sequence of compositions (given the atoms) which transforms
+ the given condition set to a set which is not a condition set i.e.,
+ FiniteSet, Interval, Set of Integers and their Union, Intersection,
+ Complement or ImageSet. We can assign a cost function to each set,
+ such that, more desirable that form of set is to us, the less the value
+ of cost function. This way our problem is now reduced to finding the path
+ from the initial ConditionSet to the lowest valued set on a graph where
+ the atomic transformations forms the edges.
+
+
+How do we deal with cases where only some of the solutions are known?
+---------------------------------------------------------------------
+
+ Creating a Universal equation solver, which can solve each and every
+ equation we encounter in mathematics is an ideal case for solvers in
+ a Computer Algebra System. We always have some cases, which are not
+ solved, or solved with incomplete solutions, so it's very important
+ to represent that situation. For this type of situation we use
+ `ConditionSet` class in the sets module, which acts as an unevaluated
+ solveset object.
```

Comment:
This might be a good place to reference [Richardson's theorem](https://en.wikipedia.org/wiki/Richardson%27s_theorem), which shows that for general classes of expressions, this is impossible.